### PR TITLE
Check if we just wrote a DotLambda before deciding on space before parens, fix 3050

### LINKED ISSUE #3050: Idempotency problem when _.Property shorthand
Issue created from [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/main?data=N4KABGBEDGD2AmBTSAuKAdSBBAQgYUzAB8A%2BMAfQDoAVWAGVgHdEAnACgEpIAacKAZ0QAXIQEsAdgHN%2BqMAG0%2BEUBBVQAJEICeAB2RpIEoT0UrIagG4BDADYBXPfJOqADL1WrIASXFJxQgMqiAF7Ibu4QACxOYAC6JgC%2BYUrRZlq6sgZ%2BxuFmVnYOCuEQAIxJ7pAAspYAHnQSiHSIUkIAFtlFxQBMztFxqokmyuUaOg6QTfAA8gBmdeKI-lrWoSkWNvayhUWdZR4Aoj4zcytFUNbTkL0JZUMeI%2Bn6AEawsNbtd3kbaFvhAMy7pm8ghYQgAYhIbAA5RCMaz1d7uIQsexXfo3VZpMbPV4IiC5dYFaKRAF4-zaSzQRA4RDTWAsRAABUsLEsAFthKxcSokSjwn0VANVLdTPcsS83gD8flNkSwABWElQMkUqk0ukNJisaCWQTecywbViWDiLkQHmIVEC9E5UUZbES1afQmnABsisgysp1Np9IAqtpdCxtbrxPrDaJjaawObLRBBSphXjbU9xVypV9HKcAOzuz2qn2IPDWHX8PDG-g86BCOlR6Y2QSxsDx5I2zF21OStbS76ygAcufJXrV9IqiFZj05irr1gbfOugwxo3bOM7TplpwAnAOVd71WXrJGp-WLXO0QvW0uUyvHQT1x1XLKPYPEFhpkJWGXWazLFGY6ereewxtleDo2muPanMUpSPnmu70v4Y6iHAB4mkeM4nu4-JxtaQGXlA9ppl2GY-O4XTbpSr7viwCGskhryHrKf6YfOQqLg8%2BEdje3aZh0-wwc%2BWAsLAtg%2BAAIogcK0VRv7IhhqhYU2OF3MBUCGIR4E8eExQRO6VTVJ40zUC0TT%2BC0dJCAA6qI8CtFGPT-thgHKXhmRGKut4QR0CqPnpBlGU0ezoaZ5lWTZbSKi69nMWerEXuxrnqR5mmkW6Pk1N40yiNUkyBpY1YsHs1TaPS-D8BGqF9lF8ksQmbFjGp7ncSRqjFDmaXVAASogcAsPAoW2YqERVSoCnNmAibqCpCWNcRsrFP27VdT18CQrY46sDMnjvqyMiKsUjZjRNqQuaytjWGIcLzKCdLfiIrDUEuM3Oh0W6PktdLwBUZ0XfU10sLd0nutALTMhSVHkIw1m2QdSkilNDVcbNpzdLpNSCSymiTCwdQVv14WVTDTlwy5CNgUlzUqJ00GnJUaMsBjWM40Iq3rSwm3bbtc2E7FuHxad52iJdiB-QD92PYjz3hJ0OyPujliY9jogVl9AtCyLeWA4%2BwOg1WrAQ1DbTc7VcX1VkT13lLfE03pABqBI4BI8ASJIeNRr2w0QKNsNJvDZsSxb7idDp7WgiJVblQ7PjO67g0e7ENUtrzptuf7nlS951s1KJsBCAA4sIhXFYgpXlTHBMOYpRM%2By5BHm2ngepTT0KwvU1JCMwTQPbo4mZeIohGuIWA%2BKOrOc6cTHVTFxtJ8uoHDBpFMQJ0bU01gcKSOIofiOHxqBBveW2PStDeL4Qh5eVtbHkbifOfFtep8lqidAtq-naw4jn%2BYGpSCPE4sD3Eh%2B7lTHkUacs5ooAR5rfMU14yZNVlJ0V61tvqiGOHQNkjx4CWCLLAMqUhm5C0vuha%2B406qz0SvA04vwHw00LqwUQ7I-A2AANKIEQNoE%2BTQhDeBwCybe%2BNThgLkiNBOpCTbkLro-FQvxqZFEgDgYs4gADWxx%2BCCWEj4aEFZECfRQULQuJUyrlhkryCBjkoHEzvpxOBSMii-BljTHAzI4KIFEkraALAGEQnfPAX0fdjTiWgMWFkA8iHgMnpA6e0CMj8x%2BvMXhFIlHCEWJoZYFDbF-CtnIlWcSqQsmgEkgISwThyI8WyXQ8BLgV0OmQ-QpN57k1lL8YONNWHsL0izP%2BMwFGWGUaoqMxRnBxy9lXSaNdrENMoXYjOciCGt2HIgHJgt6ifm0LYM%2BA8DHFyMRVceskSFHWTBxWBkyMnuF%2BI3ORdDPGMLPtYQKtF%2BACNAVfapfAYhhAMPwUEZVZBCJAPEIAA)

#### Formatted code

```fsharp
"ABC" |> _.ToLower ()

```
            

#### Reformatted code

```fsharp
"ABC" |> _.ToLower()

```
            
#### Problem description

In order for this to keep working, there can never be a space between `ToLower` and `()`.
So, the setting should not be respective in this case.

#### Extra information

- [X] The formatted result breaks my code.
- [ ] The formatted result gives compiler warnings.
- [ ] I or my company would be willing to help fix this.
- [ ] I would like a release if this problem is solved.

#### Options

Fantomas main branch at 2024-02-15T12:00:55Z - 0700fd6ec8555a4d65e11295a4a79a0fc9131f4e

```fsharp
    { config with
                SpaceBeforeUppercaseInvocation = true }
```

<sub>Did you know that you can ignore files when formatting by using a [.fantomasignore file](https://fsprojects.github.io/fantomas/docs/end-users/IgnoreFiles.html)?</sub>
<sub>PS: It's unlikely that someone else will solve your specific issue, as it's something that you have a personal stake in.</sub>

### PR DESCRIPTION
**Pull request by dawedawe**
fixes #3050

### CODE REVIEW

File: src/Fantomas.Core/Context.fs [hunk 1]
```diff
@@ -338,6 +338,18 @@ let lastWriteEventIsNewline ctx =
         | _ -> false)
     |> Option.defaultValue false
 
+let lastWriteEventIsDotLambda ctx =
```
Comment (nojaf) [inline src/Fantomas.Core/Context.fs:341]: I'm not a fan of this approach. At some point, you actually know you have a dot lambda.
By looking at the events you are just being lucky.
I would prefer to see this solved at the Oak level

**Review by nojaf — CHANGES_REQUESTED** [review]

File: src/Fantomas.Core/Context.fs (hunk 1 above)
Comment (dawedawe) [reply src/Fantomas.Core/Context.fs]: Sorry for taking that long to come back to this.
I hope this is better.

File: src/Fantomas.Core/CodePrinter.fs [hunk 2]
```diff
@@ -1553,9 +1553,24 @@ let genExpr (e: Expr) =
     | Expr.IndexFromEnd node -> !- "^" +> genExpr node.Expr |> genNode node
     | Expr.Typar node -> genSingleTextNode node
     | Expr.DotLambda node ->
+        let genDotLambdaExpr expr =
+            fun ctx ->
+                let currentConfig = ctx.Config
+
+                let dotLambdaConfig =
+                    { currentConfig with
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:1561]: Hmm, I'm a little torn about this one.
The thinking is fine, but overriding the config will open the door for the next issue to be reported:

```fsharp
[<Test>]
let ``yow`` () =
    formatSourceString
        """
let Meh () = 1

type Bar() =
    member this.Foo(v:int):int = v + 1

let b = Bar()
b |> _.Foo(Meh ())
"""
        { config with
            SpaceBeforeUppercaseInvocation = true }
    |> prepend newline
    |> should
        equal
        """
let Meh () = 1

type Bar() =
    member this.Foo(v: int) : int = v + 1

let b = Bar ()
b |> _.Foo(Meh ())
"""
```

`expr` in this case is `Meh ()` can will no longer be respected.
Checking if the `node.Expr` is parenthesis would be safer here.

Comment (dawedawe) [reply src/Fantomas.Core/CodePrinter.fs]: Thanks, much better that way.

**Review by nojaf — APPROVED** [review]
Thanks, this is good enough for now. Will probably cover the most common cases.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,10 @@
 # Changelog
 
+ ## [Unreleased]
```
Comment (nojaf) [inline CHANGELOG.md:3]: Go with a new revision for this.

**Pull request merged**